src/rgw/rgw_lc.cc

```diff
@@ -1248,32 +1252,256 @@ class LCOpAction_Transition : public LCOpAction {
+
+    pmanifest = &(*s->manifest);
+    RGWObjTier tier_config;
+    tier_config.name = oc.tier.storage_class;
+    tier_config.tier_placement = oc.tier;
+    tier_config.is_multipart_upload = tier_ctx.is_multipart_upload;
+
+    pmanifest->set_tier_type("cloud-s3");
+    pmanifest->set_tier_config(tier_config);
+
+    /* check if its necessary */
+    rgw_placement_rule target_placement;
+    target_placement.inherit_from(tier_ctx.bucket_info.placement_rule);
+    target_placement.storage_class = oc.tier.storage_class;
+    pmanifest->set_head(target_placement, (*tier_ctx.obj)->get_obj(), 0);
+
+    pmanifest->set_tail_placement(target_placement, (*tier_ctx.obj)->get_obj().bucket);
+
+    pmanifest->set_obj_size(0);
+
+    obj_op.meta.manifest = pmanifest;
+
+    /* update storage class */
+    bufferlist bl;
+    bl.append(oc.tier.storage_class);
+    attrs[RGW_ATTR_STORAGE_CLASS] = bl;
+
+    attrs.erase(RGW_ATTR_ID_TAG);
+    attrs.erase(RGW_ATTR_TAIL_TAG);
+
+    r = obj_op.write_meta(oc.dpp, 0, 0, attrs, null_yield);
+    if (r < 0) {
+      return r;
+    }
+
+    return 0;
+  }
+
+  int transition_obj_to_cloud(lc_op_ctx& oc) {
+    /* init */
+    string id = "cloudid";
+    string endpoint = oc.tier.t.s3.endpoint;
+    RGWAccessKey key = oc.tier.t.s3.key;
+    string region = oc.tier.t.s3.region;
+    HostStyle host_style = oc.tier.t.s3.host_style;
+    string bucket_name = oc.tier.t.s3.target_path;
+    const RGWZoneGroup& zonegroup = oc.store->get_zone()->get_zonegroup();
+    bool delete_object;
+   
+    /* If CurrentVersion object, remove it & create delete marker */
+    delete_object = (!oc.tier.retain_head_object ||
+                     (oc.o.is_current() && oc.bucket->versioned()));
+
+    if (bucket_name.empty()) {
+      bucket_name = "rgwx-" + zonegroup.get_name() + "-" + oc.tier.storage_class +
+                    "-cloud-bucket";
+      boost::algorithm::to_lower(bucket_name);
+    }
+
+    /* Create RGW REST connection */
+    S3RESTConn conn(oc.cct, oc.store, id, { endpoint }, key, region, host_style);
+
+    RGWLCCloudTierCtx tier_ctx(oc.cct, oc.dpp, oc.o, oc.store, oc.bucket->get_info(),
+                        &oc.obj, oc.rctx, conn, bucket_name,
+                        oc.tier.t.s3.target_storage_class);
+    tier_ctx.acl_mappings = oc.tier.t.s3.acl_mappings;
+    tier_ctx.multipart_min_part_size = oc.tier.t.s3.multipart_min_part_size;
+    tier_ctx.multipart_sync_threshold = oc.tier.t.s3.multipart_sync_threshold;
+    tier_ctx.storage_class = oc.tier.storage_class;
+
+    // check if target_path is already created
+    std::vector<std::string>& cloud_targets = oc.env.worker->get_cloud_targets();
+
+    std::vector<std::string>::iterator it = std::find(cloud_targets.begin(),
+                                                 cloud_targets.end(), bucket_name);
+    if (it != cloud_targets.end()) { //target path already created in this LC run
+      tier_ctx.create_target_bucket = false;
```

Comment:
i think std::set is a better fit than std::vector here, because it gives you O(logN) search instead of O(N)

and instead of doing the search here, and separate insert after `rgw_cloud_tier_transfer_object()` returns, you can just call `set::insert()` directly and use the `bool inserted` flag it returns:
```c++
// try to create the target bucket if this is the first time we've seen bucket_name
tier_ctx.create_target_bucket = cloud_targets.insert(bucket_name).second;
```